tracing: NPE in SELECT from crdb_internal.cluster_inflight_traces inside tenant

```
demo@127.0.0.1:57912/movr> select * from crdb_internal.cluster_inflight_traces WHERE trace_id = 4;
*
* ERROR: [sql,client=127.0.0.1:57925,hostssl,user=demo] a panic has occurred!
* runtime error: invalid memory address or nil pointer dereference
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	/usr/local/Cellar/go/1.17.2/libexec/src/runtime/panic.go:1038
*   | runtime.panicmem
*   | 	/usr/local/Cellar/go/1.17.2/libexec/src/runtime/panic.go:221
*   | runtime.sigpanic
*   | 	/usr/local/Cellar/go/1.17.2/libexec/src/runtime/signal_unix.go:735
*   | github.com/cockroachdb/cockroach/pkg/util/tracing/collector.(*TraceCollector).StartIter
*   | 	/Users/ssd/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/collector/collector.go:96
*   | github.com/cockroachdb/cockroach/pkg/sql.glob..func78
*   | 	/Users/ssd/go/src/github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:1292
*   | github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).makeConstrainedRowsGenerator.func1
*   | 	/Users/ssd/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:632
*   | github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3
*   | 	/Users/ssd/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:120
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	/Users/ssd/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:491
*   | runtime.goexit
*   | 	/usr/local/Cellar/go/1.17.2/libexec/src/runtime/asm_amd64.s:1581
* Wraps: (2) runtime error: invalid memory address or nil pointer dereference
* Error types: (1) *withstack.withStack (2) runtime.errorString
*
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x5e79ff6]

goroutine 10576 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000ae7b90, {0x9b920a8, 0xc0022d94d0})
	/Users/ssd/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:254 +0xaa
panic({0x7e23fe0, 0xc6ab760})
	/usr/local/Cellar/go/1.17.2/libexec/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/util/tracing/collector.(*TraceCollector).StartIter(0x0, {0x9b920a8, 0xc0022d9210}, 0x4)
	/Users/ssd/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/collector/collector.go:96 +0x96
github.com/cockroachdb/cockroach/pkg/sql.glob..func78({0x9b920a8, 0xc0022d9210}, {0x9d07b50, 0xc002b72f08}, 0xc003c1f8e8, {0xc002397638, 0x4682805}, 0xc004f20900)
	/Users/ssd/go/src/github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:1292 +0x105
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).makeConstrainedRowsGenerator.func1({0x9af5600, 0xc002e04e00})
	/Users/ssd/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:632 +0x4d9
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3({0x9b920a8, 0xc0022d94d0})
	/Users/ssd/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:120 +0x130
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	/Users/ssd/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:491 +0x12a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/Users/ssd/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:481 +0x445
```

--- a/pkg/ccl/logictestccl/testdata/logic_test/tenant_unsupported
+++ b/pkg/ccl/logictestccl/testdata/logic_test/tenant_unsupported
@@ -65,3 +65,8 @@ ALTER TABLE kv EXPERIMENTAL_RELOCATE LEASE VALUES (1, 'k')
 
 statement error operation is unsupported in multi-tenancy mode
 SELECT crdb_internal.check_consistency(true, '', '')
+
+# Can not query inflight traces on sql pods
+
+statement error table crdb_internal.cluster_inflight_traces is not implemented on tenants
+SELECT * from crdb_internal.cluster_inflight_traces WHERE trace_id = 42

--- a/pkg/ccl/serverccl/BUILD.bazel
+++ b/pkg/ccl/serverccl/BUILD.bazel
@@ -30,6 +30,7 @@ go_test(
         "//pkg/server",
         "//pkg/server/serverpb",
         "//pkg/sql",
+        "//pkg/sql/distsql",
         "//pkg/sql/pgwire/pgcode",
         "//pkg/sql/tests",
         "//pkg/testutils/serverutils",

--- a/pkg/ccl/serverccl/server_sql_test.go
+++ b/pkg/ccl/serverccl/server_sql_test.go
@@ -22,9 +22,11 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/security"
 	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/distsql"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/envutil"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -224,4 +226,25 @@ func TestTenantRowIDs(t *testing.T) {
 		rowCount++
 	}
 	require.Equal(t, numRows, rowCount)
+}
+
+// TestNoInflightTracesVirtualTableOnTenant verifies that internal inflight traces table
+// is correctly handled by tenants (which don't provide this functionality as of now).
+func TestNoInflightTracesVirtualTableOnTenant(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	args := base.TestClusterArgs{}
+	tc := testcluster.StartTestCluster(t, 2 /* nodes */, args)
+	defer tc.Stopper().Stop(ctx)
+
+	tenn, err := tc.Server(0).StartTenant(ctx, base.TestTenantArgs{TenantID: serverutils.TestTenantID()})
+	require.NoError(t, err, "Failed to start tenant node")
+	ex := tenn.DistSQLServer().(*distsql.ServerImpl).ServerConfig.Executor
+	_, err = ex.Exec(ctx, "get table", nil, /* txn */
+		"select * from crdb_internal.cluster_inflight_traces WHERE trace_id = 4;")
+	require.Error(t, err, "cluster_inflight_traces should be unsupported")
+	require.Contains(t, err.Error(), "table crdb_internal.cluster_inflight_traces is not implemented on tenants")
 }

--- a/pkg/sql/crdb_internal.go
+++ b/pkg/sql/crdb_internal.go
@@ -1287,6 +1287,11 @@ CREATE TABLE crdb_internal.cluster_inflight_traces (
 				"unexpected type %T for trace_id column in virtual table crdb_internal.cluster_inflight_traces", d)
 		}
 
+		if !p.ExecCfg().Codec.ForSystemTenant() {
+			// Tenant nodes doesn't have trace collector so can't serve this table.
+			return false, pgerror.New(pgcode.FeatureNotSupported,
+				"table crdb_internal.cluster_inflight_traces is not implemented on tenants")
+		}
 		traceCollector := p.ExecCfg().TraceCollector
 		var iter *collector.Iterator
 		for iter, err = traceCollector.StartIter(ctx, traceID); err == nil && iter.Valid(); iter.Next() {

